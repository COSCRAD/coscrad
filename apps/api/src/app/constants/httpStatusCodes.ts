import { ValueType } from '../../lib/types/valueType';

const httpStatusCodes = {
    ok: 200,
    badRequest: 400,
    unauthorized: 401,
    forbidden: 403,
    notFound: 404,
    internalError: 500,
};

export default httpStatusCodes;

export type HttpStatusCode = ValueType<typeof httpStatusCodes>;
