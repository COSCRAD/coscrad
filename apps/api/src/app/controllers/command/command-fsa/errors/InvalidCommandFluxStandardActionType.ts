import { InternalError } from '../../../../../lib/errors/InternalError';

export class InvalidCommandFluxStandardActionType extends InternalError {
    constructor(innerErrors: InternalError[]) {
        super(`Received a command Flux Standard Action with invalid type.`, innerErrors);
    }
}
