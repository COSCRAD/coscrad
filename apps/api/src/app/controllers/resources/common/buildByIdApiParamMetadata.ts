export default () => ({
    name: 'id',
    required: true,
    example: '2',
});
