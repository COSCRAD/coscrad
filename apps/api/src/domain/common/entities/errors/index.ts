export * from './cannot-add-duplicate-translation.error';
