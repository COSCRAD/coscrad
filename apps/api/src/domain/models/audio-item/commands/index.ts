export * from './constants';
export * from './create-audio-item';
export * from './transcripts';
