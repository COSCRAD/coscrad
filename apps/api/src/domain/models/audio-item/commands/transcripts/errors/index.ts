export * from './cannot-add-inconsistent-line-item.error';
export * from './failed-to-import-line-items-to-transcript.error';
export * from './invalid-timestamp-order.error';
export * from './invalid-transcript.error';
