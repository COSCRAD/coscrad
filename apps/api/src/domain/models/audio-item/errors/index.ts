export * from './CannotOverwriteTranscriptError';
export * from './conflicting-line-items.error';
export * from './DuplicateTranscriptParticipantErrors';
export * from './DuplicateTranscriptParticipantInitialsError';
export * from './DuplicateTranscriptParticipantNameError';
export * from './line-item-out-of-bounds.error';
export * from './transcript-line-item';
