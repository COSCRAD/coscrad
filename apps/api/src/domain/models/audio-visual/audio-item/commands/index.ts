export * from '../../shared/commands/transcripts';
export * from './constants';
export * from './create-audio-item';
export * from './translate-audio-item-name';
