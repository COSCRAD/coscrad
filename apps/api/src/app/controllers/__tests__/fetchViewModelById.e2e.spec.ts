import { CommandHandlerService } from '@coscrad/commands';
import { INestApplication } from '@nestjs/common';
import * as request from 'supertest';
import { Resource } from '../../../domain/models/resource.entity';
import { CreateSongCommandHandler } from '../../../domain/models/song/commands/create-song.command-handler';
import { PublishSongCommandHandler } from '../../../domain/models/song/commands/publish-song.command-handler';
import { InMemorySnapshotOfResources, ResourceType } from '../../../domain/types/ResourceType';
import { isInternalError } from '../../../lib/errors/InternalError';
import { ArangoConnectionProvider } from '../../../persistence/database/arango-connection.provider';
import generateRandomTestDatabaseName from '../../../persistence/repositories/__tests__/generateRandomTestDatabaseName';
import TestRepositoryProvider from '../../../persistence/repositories/__tests__/TestRepositoryProvider';
import buildTestData from '../../../test-data/buildTestData';
import httpStatusCodes from '../../constants/httpStatusCodes';
import buildMockUuidGenerator from '../command/__tests__/buildMockUuidGenerator';
import buildViewModelPathForRe from '../utilities/buildViewModelPathForResourceType';
import setUpIntegrationTest from './setUpIntegrationTest';

describe('GET /resources (fetch view models)', () => {
    const testDatabaseName = generateRandomTestDatabaseName();

    let app: INestApplication;

    let arangoConnectionProvider: ArangoConnectionProvider;

    let testRepositoryProvider: TestRepositoryProvider;

    let commandHandlerService: CommandHandlerService;

    const testData = buildTestData();

    const tagTestData = testData.tags;

    const resourceTestData = testData.resources;

    const testDataWithAllResourcesPublished = Object.entries(resourceTestData).reduce(
        (accumulatedData: InMemorySnapshotOfResources, [Re, instances]) => ({
            ...accumulatedData,
            [Re]: instances.map((instance) =>
                instance.clone({
                    published: true,
                })
            ),
        }),
        {}
    );

    beforeAll(async () => {
        ({ app, arangoConnectionProvider, testRepositoryProvider, commandHandlerService } =
            await setUpIntegrationTest({
                ARANGO_DB_NAME: testDatabaseName,
                BASE_DIGITAL_ASSET_URL: 'https://www.mysound.org/downloads/',
            }));

        commandHandlerService.registerHandler(
            'CREATE_SONG',
            new CreateSongCommandHandler(testRepositoryProvider, buildMockUuidGenerator())
        );

        commandHandlerService.registerHandler(
            'PUBLISH_SONG',
            new PublishSongCommandHandler(testRepositoryProvider, buildMockUuidGenerator())
        );
    });

    Object.values(ResourceType).forEach((resourceType) => {
        const endpointUnderTest = `/${buildViewModelPathForRe(resourceType)}`;

        const buildFullPathFromId = (id: string): string => `${endpointUnderTest}/${id}`;

        describe(`When querying for a single View Model by ID`, () => {
            beforeEach(async () => {
                await testRepositoryProvider.testSetup();
            });

            afterEach(async () => {
                await testRepositoryProvider.testTeardown();
            });
            describe(`GET ${endpointUnderTest}/:id`, () => {
                describe('when the resource is published', () => {
                    describe('when no resource with the id exists', () => {
                        beforeEach(async () => {
                            await testRepositoryProvider.addResourcesOfManyTypes(
                                testDataWithAllResourcesPublished
                            );

                            await testRepositoryProvider.getTagRepository().createMany(tagTestData);
                        });

                        it(`should return not found`, () => {
                            return request(app.getHttpServer())
                                .get(`/resources${buildFullPathFromId('bogus-id')}`)
                                .expect(httpStatusCodes.notFound);
                        });
                    });

                    describe('when an resource with the id is found', () => {
                        beforeEach(async () => {
                            await testRepositoryProvider.addResourcesOfManyTypes(
                                testDataWithAllResourcesPublished
                            );

                            await testRepositoryProvider.getTagRepository().createMany(tagTestData);
                        });

                        it('should return the expected response', async () => {
                            const resourceToFind =
                                testDataWithAllResourcesPublished[resourceType][0];

                            const res = await request(app.getHttpServer()).get(
                                `/resources${buildFullPathFromId(resourceToFind.id)}`
                            );

                            expect(res.status).toBe(httpStatusCodes.ok);

                            expect(res.body.data.id).toBe(resourceToFind.id);

                            expect(res.body).toMatchSnapshot();
                        });
                    });
                });

                describe('when an resource with the id is unpublished', () => {
                    const unpublishedId = 'unpublished-01';

                    beforeEach(async () => {
                        await testRepositoryProvider.addResourcesOfManyTypes(resourceTestData);

                        const unpublishedInstance = resourceTestData[resourceType][0].clone({
                            published: false,
                            id: unpublishedId,
                        });

                        await testRepositoryProvider.addResourcesOfSingleType(resourceType, [
                            unpublishedInstance,
                        ]);
                    });

                    it('should return not found', async () => {
                        const publishedAndUnpublishedInstancesFromRepo =
                            await testRepositoryProvider
                                .forResource(resourceType)
                                .fetchMany()
                                .then((result) =>
                                    result.filter(
                                        (singleInstance): singleInstance is Resource =>
                                            !isInternalError(singleInstance)
                                    )
                                );

                        /**
                         * Given 404 is not a very specific symptom, let's be sure the
                         * unpubished resource was in the db to start with
                         */
                        const isUnpublishedresourceIdInDB =
                            publishedAndUnpublishedInstancesFromRepo.some(
                                ({ id }) => id === unpublishedId
                            );

                        expect(isUnpublishedresourceIdInDB).toBe(true);

                        return request(app.getHttpServer())
                            .get(`/resources${buildFullPathFromId(unpublishedId)}`)
                            .expect(httpStatusCodes.notFound);
                    });
                });
            });
        });
    });

    afterAll(async () => {
        await arangoConnectionProvider.dropDatabaseIfExists();

        await app.close();
    });
});
