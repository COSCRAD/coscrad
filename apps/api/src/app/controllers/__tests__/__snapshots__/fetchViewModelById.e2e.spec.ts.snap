// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET  (fetch view models) When querying for a single View Model by ID GET /resources/audioItems/:id when the resource is published when an resource with the id exists should return the expected response 1`] = `
{
  "accessControlList": {
    "allowedGroupIds": [],
    "allowedUserIds": [],
  },
  "actions": [],
  "audioURL": "https://www.testapi.coscrad.org/api/resources/mediaItems/download/9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
  "contributions": [],
  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110110",
  "isPublished": false,
  "lengthMilliseconds": 20000,
  "mimeType": "audio/wav",
  "name": {
    "items": [
      {
        "languageCode": "en",
        "role": "original",
        "text": "The Wooden Boy",
      },
    ],
  },
  "tags": [
    {
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110003",
      "label": "placenames",
      "members": [
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110000",
          "type": "photograph",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110110",
          "type": "audioItem",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b112003",
          "type": "note",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110005",
          "type": "note",
        },
      ],
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "placenames",
          },
        ],
      },
    },
  ],
  "text": "[12000] [JHC] There once was a little wooden boy. {en} (role: original) [15550]
[18300] [JHC] His name was Pinocchio {en} (role: original) [19240]",
}
`;

exports[`GET  (fetch view models) When querying for a single View Model by ID GET /resources/bibliographicCitations/:id when the resource is published when an resource with the id exists should return the expected response 1`] = `
{
  "actions": [],
  "contributions": [],
  "data": {
    "abstract": "This is the abstract, NOT a general note!",
    "creators": [
      {
        "name": "Alana Duvernay",
        "type": "author",
      },
      {
        "name": "James Smith",
        "type": "author",
      },
    ],
    "isbn": "978-1-895811-34-6",
    "numberOfPages": 455,
    "place": "Kennebunk Port, Maine",
    "publisher": "Atlantic Publishing",
    "title": "A Day in the Life",
    "type": "book",
    "url": "https://coscrad.org/wp-content/uploads/2023/05/mock-book-bibliographic-citation-1_a-day-in-the-life.pdf",
    "year": 1999,
  },
  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
  "name": {
    "items": [
      {
        "languageCode": "en",
        "role": "original",
        "text": "A Day in the Life",
      },
    ],
  },
  "tags": [
    {
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110002",
      "label": "animals",
      "members": [
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110002",
          "type": "term",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110003",
          "type": "term",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110102",
          "type": "spatialFeature",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b112004",
          "type": "note",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110007",
          "type": "note",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
          "type": "bibliographicCitation",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110223",
          "type": "video",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110951",
          "type": "digitalText",
        },
      ],
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "animals",
          },
        ],
      },
    },
  ],
}
`;

exports[`GET  (fetch view models) When querying for a single View Model by ID GET /resources/mediaItems/:id when the resource is published when an resource with the id exists should return the expected response 1`] = `
{
  "actions": [],
  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110009",
  "mimeType": "audio/wav",
  "name": {
    "items": [
      {
        "languageCode": "clc",
        "role": "original",
        "text": "No Light",
      },
    ],
  },
  "tags": [],
  "url": "https://www.testapi.coscrad.org/api/resources/mediaItems/download/9b1deb4d-3b7d-4bad-9bdd-2b0d7b110009",
}
`;

exports[`GET  (fetch view models) When querying for a single View Model by ID GET /resources/playlists/:id when the resource is published when an resource with the id exists should return the expected response 1`] = `
{
  "actions": [],
  "contributions": [],
  "episodes": [
    {
      "mediaItemUrl": "/resources/mediaItems/download/9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
      "mimeType": "audio/wav",
      "name": "The Wooden Boy {en} (role: original)",
    },
    {
      "mediaItemUrl": "/resources/mediaItems/download/9b1deb4d-3b7d-4bad-9bdd-2b0d7b110003",
      "mimeType": "audio/mpeg",
      "name": "Down at the River {clc} (role: original)",
    },
  ],
  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110501",
  "name": {
    "items": [
      {
        "languageCode": "clc",
        "role": "original",
        "text": "Blake's jams",
      },
    ],
  },
  "tags": [
    {
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110006",
      "label": "law",
      "members": [
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110003",
          "type": "bibliographicCitation",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110501",
          "type": "playlist",
        },
      ],
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "law",
          },
        ],
      },
    },
  ],
}
`;

exports[`GET  (fetch view models) When querying for a single View Model by ID GET /resources/spatialFeatures/:id when the resource is published when an resource with the id exists should return the expected response 1`] = `
{
  "actions": [],
  "geometry": {
    "coordinates": [
      54.034404852745006,
      -132.17563335558532,
    ],
    "type": "Point",
  },
  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110100",
  "name": {
    "items": [
      {
        "languageCode": "en",
        "role": "original",
        "text": "Name of Point with ID: 100",
      },
    ],
  },
  "properties": {
    "description": "Description for point 100",
    "imageUrl": "https://coscrad.org/wp-content/uploads/2023/05/evergreen-2025158_1280.png",
    "name": "Name of Point with ID: 100",
  },
  "tags": [
    {
      "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
      "label": "plants",
      "members": [
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
          "type": "term",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110001",
          "type": "vocabularyList",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110100",
          "type": "spatialFeature",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b112005",
          "type": "note",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110009",
          "type": "note",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110011",
          "type": "note",
        },
        {
          "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110950",
          "type": "digitalText",
        },
      ],
      "name": {
        "items": [
          {
            "languageCode": "en",
            "role": "original",
            "text": "plants",
          },
        ],
      },
    },
  ],
}
`;

exports[`GET  (fetch view models) When querying for a single View Model by ID GET /resources/videos/:id when the resource is published when an resource with the id exists should return the expected response 1`] = `
{
  "actions": [],
  "contributions": [],
  "id": "9b1deb4d-3b7d-4bad-9bdd-2b0d7b110223",
  "lengthMilliseconds": 20000,
  "mimeType": "video/mp4",
  "name": {
    "items": [
      {
        "languageCode": "en",
        "role": "original",
        "text": "The Demonstration",
      },
    ],
  },
  "tags": [],
  "transcript": {
    "items": [
      {
        "inPointMilliseconds": 12000,
        "outPointMilliseconds": 15550,
        "speakerInitials": "DM",
        "text": {
          "items": [
            {
              "languageCode": "en",
              "role": "original",
              "text": "This is how.",
            },
          ],
        },
      },
      {
        "inPointMilliseconds": 18300,
        "outPointMilliseconds": 19240,
        "speakerInitials": "DM",
        "text": {
          "items": [
            {
              "languageCode": "en",
              "role": "original",
              "text": "It is done",
            },
          ],
        },
      },
    ],
    "participants": [
      {
        "initials": "DM",
        "name": "Dee Monstrator",
      },
    ],
  },
  "videoUrl": "/resources/mediaItems/download/9b1deb4d-3b7d-4bad-9bdd-2b0d7b110002",
}
`;
